packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java

```diff
@@ -5,42 +5,204 @@
 package io.flutter.plugins.sharedpreferences;
 
 import android.content.Context;
+import android.content.SharedPreferences;
+import android.util.Base64;
+import android.util.Log;
+import androidx.annotation.NonNull;
+import androidx.annotation.VisibleForTesting;
 import io.flutter.embedding.engine.plugins.FlutterPlugin;
 import io.flutter.plugin.common.BinaryMessenger;
-import io.flutter.plugin.common.MethodChannel;
+import io.flutter.plugins.sharedpreferences.Messages.SharedPreferencesApi;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
 
 /** SharedPreferencesPlugin */
-public class SharedPreferencesPlugin implements FlutterPlugin {
-  private static final String CHANNEL_NAME = "plugins.flutter.io/shared_preferences_android";
-  private MethodChannel channel;
-  private MethodCallHandlerImpl handler;
+public class SharedPreferencesPlugin implements FlutterPlugin, SharedPreferencesApi {
+  private static final String TAG = "SharedPreferencesPlugin";
+  private static final String SHARED_PREFERENCES_NAME = "FlutterSharedPreferences";
+  private static final String LIST_IDENTIFIER = "VGhpcyBpcyB0aGUgcHJlZml4IGZvciBhIGxpc3Qu";
+  private static final String BIG_INTEGER_PREFIX = "VGhpcyBpcyB0aGUgcHJlZml4IGZvciBCaWdJbnRlZ2Vy";
+  private static final String DOUBLE_PREFIX = "VGhpcyBpcyB0aGUgcHJlZml4IGZvciBEb3VibGUu";
+
+  private SharedPreferences preferences;
+  private SharedPreferencesListEncoder listEncoder;
+
+  public SharedPreferencesPlugin() {
+    this(new ListEncoder());
+  }
+
+  @VisibleForTesting
+  SharedPreferencesPlugin(SharedPreferencesListEncoder testListEncoder) {
```

Comment:
Nit: `testListEncoder` is misleading since it's the real encoder when called from the other constructor. You should be able to just call it `listEncoder` then do `this.listEncoder = listEncoder;`